source/common/ssl/ssl_socket.cc

```diff
@@ -376,10 +376,12 @@ std::string SslSocket::subjectLocalCertificate() const {
 ClientSslSocketFactory::ClientSslSocketFactory(const ClientContextConfig& config,
                                                Ssl::ContextManager& manager,
                                                Stats::Scope& stats_scope)
-    : ssl_ctx_(manager.createSslClientContext(stats_scope, config)) {}
+    : manager_(manager), ssl_ctx_(manager.createSslClientContext(stats_scope, config)) {}
+
+ClientSslSocketFactory::~ClientSslSocketFactory() { manager_.removeContext(ssl_ctx_.get()); }
```

Comment:
in `~ContextImpl` it already calls `removeContext`, isn't this removing twice? since `remove` doesn't raise exceptions when there is nothing to remove, you may want to add an `ASSERT` in `removeContext`.

Reply:
Good point.  It was added in the final PR with secret update code to fix some integration test failures.  I reverted the changes for now.